pubsub/fakepubsub/fake.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fakepubsub provides an in-memory fake pubsub implementation.
+// This should not be used for production: it is intended for local
+// development.
+//
+// fakepubsub does not support any types for As.
+package fakepubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/google/go-cloud/pubsub/driver"
+)
+
+type topic struct {
+	mu        sync.Mutex
+	subs      []*subscription
+	nextAckID int
+	closed    bool
+}
+
+// OpenTopic establishes a new topic.
+// Open subscribers for the topic before publishing.
+func OpenTopic() driver.Topic {
+	return &topic{}
+}
+
+// SendBatch implements driver.SendBatch.
+// It is error if the topic is closed or has no subscriptions.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// Check for closed or canceled before doing any work.
+	if t.closed {
+		return errors.New("fakepubsub: SendBatch: topic closed")
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	// Catch the common mistake of publishing before any subscriptions have been created.
+	if len(t.subs) == 0 {
+		return errors.New("fakepubsub: SendBatch: no subscriptions for topic")
+	}
+	// Associate ack IDs with messages here. It would be a bit better if each subscription's
+	// messages had their own ack IDs, so we could catch one subscription using ack IDs from another,
+	// but that would require copying all the messages.
+	for i, m := range ms {
+		m.AckID = t.nextAckID + i
+	}
+	t.nextAckID += len(ms)
+	for _, s := range t.subs {
+		s.add(ms)
+	}
+	return nil
+}
+
+// Close closes the topic. Subsequent calls to SendBatch will fail.
+func (t *topic) Close() error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.closed = true
```

Comment:
Shouldn't `Close` block until any in-flight `SendBatch` finish?

Reply:
The mutex should take care of that, no?

Reply:
Yes, you're right.